### PR TITLE
Add info on UI 'play.crypto.provider' exception

### PR DESCRIPTION
**Pull request by AlexDBlack**

### CODE REVIEW

**Comment by linghongtao** [comment]
if use shade plugin ，after mvn install the jar is only 1068kb，but my project have spark and deeplearning4j ,the jar will be 342M when I use assembly plugin

**Comment by linghongtao** [comment]
my code is like this 

1. `package dl4jtest




import java.io.File
import java.util
import java.util.{Collections, Random}

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import org.deeplearning4j.eval.Evaluation
import org.deeplearning4j.nn.api.OptimizationAlgorithm
import org.deeplearning4j.nn.conf.{BackpropType, MultiLayerConfiguration, NeuralNetConfiguration, Updater}
import org.deeplearning4j.nn.conf.layers.{GravesLSTM, RnnOutputLayer}
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork
import org.deeplearning4j.nn.weights.WeightInit
import org.deeplearning4j.optimize.listeners.ScoreIterationListener
import org.deeplearning4j.spark.api.RDDTrainingApproach
import org.deeplearning4j.spark.impl.multilayer.SparkDl4jMultiLayer
import org.deeplearning4j.spark.impl.paramavg.ParameterAveragingTrainingMaster
import org.deeplearning4j.util.ModelSerializer
import org.nd4j.linalg.activations.Activation
import org.nd4j.linalg.dataset.DataSet
import org.nd4j.linalg.factory.Nd4j
import org.nd4j.linalg.lossfunctions.LossFunctions.LossFunction
import org.slf4j.{Logger, LoggerFactory}
import org.deeplearning4j.ui.stats.StatsListener
import org.deeplearning4j.ui.storage.FileStatsStorage
import java.util.Collections

import org.apache.spark.storage.StorageLevel
import org.deeplearning4j.ui.api.UIServer

import scala.collection.mutable.ArrayBuffer
object SparkLstmScala {
  private val log: Logger = LoggerFactory.getLogger(SparkLstmScala.getClass)
  def main(args: Array[String]): Unit = {
    if(args.length<1){
      """
        |
      """.stripMargin
      sys.exit(0)
    }
    val Array(date)=args

    val spark=SparkSession.builder()
      .appName("SparkLstmScala")
      .config("spark.serializer","org.apache.spark.serializer.KryoSerializer")
      .config("spark.kryo.registrator", "org.nd4j.Nd4jRegistrator")
        //.master("local[*]")
      //.enableHiveSupport()
      .getOrCreate()

    val trainrdd: RDD[DataSet] =spark.read.text("hdfs://ns-bj/project/vine/lstmrdd/datadistinct/"+date+"/").rdd
      .map(row => row.getString(0))
      .map(line=>{
        val array = line.split('"')
        val sid = array(0).split("-")
        val sidarr=sid(sid.size-1).split(",")
        val feature = array(3)
          .replaceAll("ArrayBuffer\\(", "")
          .replaceAll("\\)", "")
          .replaceAll(" ", "")
        val label = array(4).replaceAll(",", "")

        (sidarr(0),(sidarr(1),feature,label))
      }).sortBy(_._1)
      .map(row=>(row._2._1,(row._2._2,row._2._3))).groupByKey()
      .flatMap(iter => toDataSet(iter))
      .persist(StorageLevel.MEMORY_AND_DISK)

    val weights = Array[Double](0.7,0.3)
    val seed:Long=123456
    val Array(traindata,testdata)=trainrdd.randomSplit(weights,seed)

    //trainrdd.take(100).foreach(println(_))

    val rng: Random = new Random(12345)
    //Number of units in each GravesLSTM layer
    val lstmLayerSize = 200
    //Length for truncated backpropagation through time. i.e., do parameter updates ever 50 characters
    val tbpttLength = 50
    //Number of samples to generate after each training epoch
    val nSamplesToGenerate = 4
    //Optional character initialization; a random character is used if null
    //String generationInitialization = null;

    //Set up network configuration:
    val conf: MultiLayerConfiguration = new NeuralNetConfiguration.Builder()
      .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
      .iterations(1)
      .learningRate(0.1)
      .seed(12345)
      .regularization(true)
      .l2(0.001)
      .weightInit(WeightInit.XAVIER)
      .updater(Updater.RMSPROP)   //To configure: .updater(new RmsProp(0.95))
      .list()
      .layer(0, new GravesLSTM.Builder()
        .nIn(FEATURE_DIMS)
        .nOut(lstmLayerSize)
        .activation(Activation.TANH).build())
      .layer(1, new GravesLSTM.Builder()
        .nIn(lstmLayerSize)
        .nOut(lstmLayerSize)
        .activation(Activation.TANH).build())
      .layer(2, new RnnOutputLayer.Builder(LossFunction.MCXENT)
        .nIn(lstmLayerSize)
        .nOut(1)
        .activation(Activation.SOFTMAX).build())//MCXENT + softmax for classification
      .backpropType(BackpropType.TruncatedBPTT)
      .tBPTTForwardLength(tbpttLength)
      .tBPTTBackwardLength(tbpttLength)
      .pretrain(false)
      .backprop(true)
      .build()

    val batchSizePerWorker = 5
    val averagingFrequency = 5

    val tm: ParameterAveragingTrainingMaster = new ParameterAveragingTrainingMaster
    .Builder(dataSetObjectSize)
      .workerPrefetchNumBatches(2)//Asynchronously prefetch up to 2 batches
      .averagingFrequency(averagingFrequency)
      .batchSizePerWorker(batchSizePerWorker)
      .rddTrainingApproach(RDDTrainingApproach.Direct)
      //.exportDirectory("D:/sparkDataOut/lstm/export")
      .build()

    val sparkNetwork: SparkDl4jMultiLayer = new SparkDl4jMultiLayer(spark.sparkContext, conf, tm)
    sparkNetwork.setListeners(new ScoreIterationListener(1))

    val uiserver = UIServer.getInstance()
    uiserver.enableRemoteListener()

    import org.deeplearning4j.api.storage.impl.RemoteUIStatsStorageRouter
    val router = new RemoteUIStatsStorageRouter("http://localhost:9000")

   sparkNetwork.setListeners(router,Collections.singletonList(new StatsListener(null)))

    val numEpochs = 10
    var net: MultiLayerNetwork = null
    //Do training, and then generate and print samples from network
    for (i <- 0 until numEpochs) {
      //Perform one epoch of training. At the end of each epoch, we are returned a copy of the trained network
      sparkNetwork.fit(traindata)
    }




    val strings = new util.ArrayList[String]()
    val evaluation = sparkNetwork.doEvaluation(testdata, 64, new Evaluation(1))(0) //Work-around for 0.9.1 bug: see https://deeplearning4j.org/releasenotes
    log.info("***** Evaluation *****")
    log.info(evaluation.stats)
    val stats: String = evaluation.stats
    strings.add(stats)


    val network: MultiLayerNetwork = sparkNetwork.getNetwork
    val str: String = network.summary()
    println(str)

    strings.add(str)

    /*spark
      .sparkContext
      .parallelize(strings.toArray)
      .repartition(1)
      .saveAsTextFile("D:/sparkDataOut/lstmoutput")*/

//    val model: MultiLayerNetwork = new MultiLayerNetwork(conf)
   /* ModelSerializer.writeModel(sparkNetwork.getNetwork,
      "D:/sparkDataOut/lstmModel",false)*/

   /* val nn = ModelSerializer.restoreMultiLayerNetwork("D:/sparkDataOut/lstmModel")
    val ss = new SparkDl4jMultiLayer(spark.sparkContext, nn, tm)*/


    //uiserver.stop()
    spark.stop()


  }
  val dataSetObjectSize = 32 //how many samples for each dataset
  val FEATURE_DIMS = 115
  val SEQ_NO = 32

  def toDataSet(iter: (String, Iterable[(String, String)])): Iterator[DataSet] ={
    val ite: scala.Iterator[Seq[(String, String)]]#GroupedIterator[Seq[(String, String)]] = iter._2.iterator
      .sliding(SEQ_NO, SEQ_NO)
      .filter(_.size == SEQ_NO)
      .sliding(dataSetObjectSize, dataSetObjectSize)

    val buf = new ArrayBuffer[DataSet]

    while (ite.hasNext){
      val SampleList: scala.Iterator[Seq[(String, String)]] = ite.next().iterator
      val features = Nd4j.zeros(dataSetObjectSize,FEATURE_DIMS,SEQ_NO)
      val labels = Nd4j.zeros(dataSetObjectSize,1,SEQ_NO)

      var sampleIndex=0

      while(SampleList.hasNext){
        val sample: Seq[(String, String)] = SampleList.next()

        for(seqNo<-0 until SEQ_NO){
          val (feature,label): (String, String) = sample.apply(seqNo)
          val dims = feature.split(",").map(_.toDouble)

          for(dimIndex<-0 until FEATURE_DIMS){
            features.putScalar(sampleIndex,dimIndex,seqNo,dims(dimIndex))
          }
          labels.putScalar(sampleIndex,1,seqNo,label.toDouble)
        }
          sampleIndex=sampleIndex + 1
      }
      buf+=new DataSet(features,labels)
    }
    buf.iterator
  }
}
`

and the exception is same as you 

my pom.xml is 

` <dependency>
            <groupId>org.nd4j</groupId>
            <artifactId>nd4j-native-platform</artifactId>
            <version>0.9.1</version>
        </dependency>

        <dependency>
            <groupId>org.nd4j</groupId>
            <artifactId>nd4j-kryo_${scala.binary.version}</artifactId>
            <version>0.9.1</version>
        </dependency>
        
        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>dl4j-spark_${scala.binary.version}</artifactId>
            <version>0.9.1_spark_2</version>
            <exclusions>
                <exclusion>
                    
                    <groupId>org.bytedeco.javacpp-presets</groupId>
                    <artifactId>opencv</artifactId>
                </exclusion>
            </exclusions>
        </dependency>

        
        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-ui_${scala.binary.version}</artifactId>
            <version>0.9.1</version>
        </dependency>`

what is wrong with me??????????????


please help?

**Comment by agibsonccc** [comment]
You can't use the UI + spark together in the same uber jar. Start it as a separate process.

**Comment by linghongtao** [comment]
how to Start it as a separate process?  could you give me a example ,?Our task must be run in the spark ,we have no choice

**Comment by AlexDBlack** [comment]
@linghongtao see here: https://deeplearning4j.org/visualization#sparkui

**Comment by agibsonccc** [comment]
It's called reading our docs: http://deeplearning4j.org/visualization - we cover this to death in our examples. http://github.com/deeplearning4j/dl4j-examples - please put some effort in. We are not a compiler. I am locking this  issue . If you want further help (hint: do *not* come in to the channel without having read the docs first) we will help you more here: https://gitter.im/deeplearning4j/deeplearning4j

**Comment by linghongtao** [comment]
@AlexDBlack @agibsonccc @huitseeker @mneedham @joebowbeer ,Our team is iflytek advertise ,we like use deeplearning4J ,we think if DL4J could merge spark more easily,more perfect ,DL4J can kill tensorflow,
because tensorflow can not run on spark well.thank

**Comment by agibsonccc** [comment]
Great! I'm assuming your team is in china, please feel free to join our qq or the chinese gitter. We always like hearing from users.

**Comment by linghongtao** [comment]
what is your qq ,we are china enterprise ifytek -----you can see our website http://www.iflytek.com/en/ 
thank you.

we hope DL4J and spark could be merge. It will be standard in the dataming(except image and text)

we need one deeplearning framework is just use in data and could run well on spark .thank you

**Comment by linghongtao** [comment]
if Dl4j could run well on spark and focus on data only ,there well be more and more users ,because now ,our data all storage on hadoop and compute by spark

**Comment by linghongtao** [comment]
@agibsonccc  what is qq of yours?

**Comment by agibsonccc** [comment]
@likeliberation look at the bottom: https://deeplearning4j.org/cn/
